### PR TITLE
Adjoint batching support to `Adjoint` wrapper

### PR DESCRIPTION
**Pull request by albi3ro**
The matrix for adjoint now works with batched parameters:
```
>>> base = qml.RX(np.array([1.2, 2.3]), 0)
>>> op = qml.adjoint(base, lazy=True)
>>> op.matrix()
tensor([[[0.82533561-0.j        , 0.        +0.56464247j],
         [0.        +0.56464247j, 0.82533561-0.j        ]],

        [[0.40848744-0.j        , 0.        +0.91276394j],
         [0.        +0.91276394j, 0.40848744-0.j        ]]], requires_grad=True)
```

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -49,6 +49,8 @@ keyword argument when using `GellMann`, which determines which of the 8 Gell-Man
 
 <h3>Improvements</h3>
 
+* `Adjoint` now supports batching if the base operation supports batching.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:52]:
```suggestion
* `Adjoint` now supports batching if the base operation supports batching.
  [(#3168)](https://github.com/PennyLaneAI/pennylane/pull/3168)
```

File: pennylane/ops/op_math/adjoint_class.py [hunk 2]
```diff
@@ -204,26 +204,36 @@ def __init__(self, base=None, do_queue=True, id=None):
     def __repr__(self):
         return f"Adjoint({self.base})"
 
+    # pylint: disable=protected-access
+    def _check_batching(self, params):
+        self.base._check_batching(params)
```
Comment (dwierichs) [inline pennylane/ops/op_math/adjoint_class.py:209]: As `SymbolicOp.__init__` does not inherit from `Operator.__init__`, `Adjoint._check_batching` is not being called anywhere, as far as I am aware. Maybe I am missing something here?
My suggestion would be to leave the definitions of `batch_size` and `ndim_params` from below in, but simply skip this one.

File: tests/ops/op_math/test_adjoint_op.py [hunk 3]
```diff
@@ -308,6 +308,17 @@ class DummyOp(qml.operation.Operator):
...
+    def test_batching_properties(self):
+        """Test the batching properties and methods."""
+
+        base = qml.RX(np.array([1.2, 2.3, 3.4]), 0)
+        op = Adjoint(base)
+        assert op.batch_size == 3
+        assert op.ndim_params == (0,)
+
+        op._check_batching([np.array([1.2, 2.3])])
+        assert op.batch_size == base.batch_size == 2
```
Comment (dwierichs) [inline tests/ops/op_math/test_adjoint_op.py:320]: This behaviour makes sense if we see `Adjoint(op)` to "own" `op`. :+1: 
However, I'm not aware of any scenario where `Adjoint._check_batching` is called, so this test might be somewhat in void.

File: tests/ops/op_math/test_adjoint_op.py [hunk 4]
```diff
@@ -602,14 +613,16 @@ def test_do_queue_False(self):
...
         x = qml.numpy.array([0.1, 0.2, 0.3])
         base = qml.RX(x, wires=0)
         op = Adjoint(base)
+        mat = op.matrix()
+        compare = qml.RX(-x, wires=0)
 
-        with pytest.raises(qml.operation.MatrixUndefinedError):
-            op.matrix()
+        assert qml.math.allclose(mat, compare.matrix())
+        assert mat.shape == (3, 2, 2)
```
Comment (dwierichs) [inline tests/ops/op_math/test_adjoint_op.py:625]: :+1:

**Review by dwierichs — COMMENTED** [review]
Looks good to me @albi3ro, except for the definition (and test) of `_check_batching` which might be superfluous I think.

File: pennylane/ops/op_math/adjoint_class.py [hunk 5]
```diff
@@ -204,26 +204,36 @@ def __init__(self, base=None, do_queue=True, id=None):
...
         else:
             base_matrix = self.base.matrix(wire_order=wire_order)
 
-        return transpose(conj(base_matrix))
+        return moveaxis(conj(base_matrix), -2, -1)
+
+    # pylint: disable=arguments-differ
+    def sparse_matrix(self, wire_order=None, format="csr"):
+        base_matrix = self.base.sparse_matrix(wire_order=wire_order)
+        return transpose(conj(base_matrix)).asformat(format=format)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/op_math/adjoint_class.py:234]: I really don't want to open the can of worms that is batched sparse matrices. Sparse matrices are two dimensional only.

File: tests/ops/op_math/test_adjoint_op.py (hunk 3 above)
Comment (albi3ro) [reply tests/ops/op_math/test_adjoint_op.py:320]: Updating the data of `Adjoint`?  
```python
@qml.qnode(qml.device('default.qubit', wires=1), diff_method="parameter-shift")
def circuit(x):
    qml.adjoint(qml.RX(x, wires=0))
    return qml.expval(qml.PauliZ(0))

qml.grad(circuit)(qml.numpy.array(0.1))
```
Gets called a bunch.

File: pennylane/ops/op_math/adjoint_class.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/ops/op_math/adjoint_class.py:209]: Oh I see, I was indeed missing something.

I'm definitely in favour of implementing a property setter. I think the operator algorithmic work is really highlighting edge cases in the sense that it makes use of class features standard ops don't use, and this should definitely be made clearer!

I'm wondering: Are you thinking of an approach that would make it unnecessary for developers to think about the `batch_size` updating at all? :thinking:

File: tests/ops/op_math/test_adjoint_op.py (hunk 3 above)
Comment (dwierichs) [reply tests/ops/op_math/test_adjoint_op.py:320]: Yeah, this is the result of me missing the updating above :) Thanks!

**Review by dwierichs — APPROVED** [review]
My questions are answered, thank you @albi3ro!
Good to go from my side.

File: pennylane/ops/op_math/adjoint_class.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/ops/op_math/adjoint_class.py:209]: It'd be great if someone could just do:
```
op.data = new_data
```
and have batch size automatically updated.  But we can discuss this later elsewhere.

**Pull request merged**